iceoryx_dds/source/iceoryx_dds/dds/fastdds_data_writer.cpp

```diff
@@ -0,0 +1,165 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "iceoryx_dds/dds/fastdds_data_writer.hpp"
+#include "iceoryx_dds/internal/log/logging.hpp"
+#include "MempoolPubSubTypes.h"
+
+#include <fastdds/dds/domain/DomainParticipantFactory.hpp>
+#include <fastdds/dds/publisher/qos/DataWriterQos.hpp>
+#include <string>
+
+iox::dds::FastDDSDataWriter::FastDDSDataWriter(
+        iox::capro::IdString_t serviceId,
+        iox::capro::IdString_t instanceId,
+        iox::capro::IdString_t eventId)
+    : m_serviceId(serviceId)
+    , m_instanceId(instanceId)
+    , m_eventId(eventId)
+    , m_type(new Mempool::ChunkPubSubType())
+    , m_listener(nullptr)
+{
+    m_listener.fastddsDataWriter = this;
+    LogDebug() << m_logName << "Created FastDDSDataWriter";
+}
+
+iox::dds::FastDDSDataWriter::~FastDDSDataWriter()
+{
+    if (m_dataWriter != nullptr)
+    {
+        m_publisher->delete_datawriter(m_dataWriter);
+    }
+    if (m_topic != nullptr)
+    {
+        m_participant->delete_topic(m_topic);
+    }
+    if (m_publisher != nullptr)
+    {
+        m_participant->delete_publisher(m_publisher);
+    }
+    if (m_participant != nullptr)
+    {
+        eprosima::fastdds::dds::DomainParticipantFactory::get_instance()->delete_participant(m_participant);
+    }
+    LogDebug() << m_logName << "Destroyed FastDDSDataWriter";
+}
+
+void iox::dds::FastDDSDataWriter::connect() noexcept
+{
+    if (!m_isConnected.load(std::memory_order_relaxed))
+    {
+        // Create the DomainParticipant
+        eprosima::fastdds::dds::DomainParticipantQos participantQos;
+        m_participant =
+            eprosima::fastdds::dds::DomainParticipantFactory::get_instance()->create_participant(0, participantQos);
+
+        // Register the type
+        m_type.register_type(m_participant);
+
+        // Create topic
+        auto m_topicname =
+            "/" + std::string(m_serviceId) + "/" + std::string(m_instanceId) + "/" + std::string(m_eventId);
+        m_topic = m_participant->create_topic(m_topicname, "Mempool::Chunk", eprosima::fastdds::dds::TOPIC_QOS_DEFAULT);
+
+        // Create publisher
+        m_publisher = m_participant->create_publisher(eprosima::fastdds::dds::PUBLISHER_QOS_DEFAULT, nullptr);
+
+        // Create datawriter
+        eprosima::fastdds::dds::DataWriterQos dataWriterQos = eprosima::fastdds::dds::DATAWRITER_QOS_DEFAULT;
+        dataWriterQos.endpoint().history_memory_policy =
+            eprosima::fastrtps::rtps::PREALLOCATED_WITH_REALLOC_MEMORY_MODE;
+        m_dataWriter = m_publisher->create_datawriter(m_topic, dataWriterQos, &m_listener);
+
+        // Save state
+        m_isConnected.store(true, std::memory_order_relaxed);
```

Comment:
We may have here also a memory synchronization problem. See `fastdds_data_reader.cpp` comment in the connect method.